ff-fft/src/domain/radix2.rs

```diff
@@ -0,0 +1,312 @@
+//! performing various kinds of polynomial arithmetic on top of
+//! the scalar field using a radix-2 domain.
+
+pub use crate::domain::utils::Elements;
+use crate::domain::{
+    utils::{best_fft, bitreverse},
+    DomainCoeff, EvaluationDomain,
+};
+use crate::Vec;
+use algebra_core::{FftField, FftParameters};
+use core::fmt;
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+/// Defines a domain over which finite field (I)FFTs can be performed. Works
+/// only for fields that have a large multiplicative subgroup of size that is
+/// a power-of-2.
+#[derive(Copy, Clone, Hash, Eq, PartialEq)]
+pub struct Radix2EvaluationDomain<F: FftField> {
+    /// The size of the domain.
+    pub size: u64,
+    /// `log_2(self.size)`.
+    pub log_size_of_group: u32,
+    /// Size of the domain as a field element.
+    pub size_as_field_element: F,
+    /// Inverse of the size in the field.
+    pub size_inv: F,
+    /// A generator of the subgroup.
+    pub group_gen: F,
+    /// Inverse of the generator of the subgroup.
+    pub group_gen_inv: F,
+    /// Multiplicative generator of the finite field.
+    pub generator_inv: F,
+}
+
+impl<F: FftField> fmt::Debug for Radix2EvaluationDomain<F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Radix-2 multiplicative subgroup of size {}", self.size)
+    }
+}
+
+impl<F: FftField> EvaluationDomain<F> for Radix2EvaluationDomain<F> {
+    type Elements = Elements<F>;
+
+    /// Construct a domain that is large enough for evaluations of a polynomial
+    /// having `num_coeffs` coefficients.
+    fn new(num_coeffs: usize) -> Option<Self> {
+        // Compute the size of our evaluation domain
+        let size = num_coeffs.next_power_of_two() as u64;
+        let log_size_of_group = size.trailing_zeros();
+
+        // libfqfft uses > https://github.com/scipr-lab/libfqfft/blob/e0183b2cef7d4c5deb21a6eaf3fe3b586d738fe0/libfqfft/evaluation_domain/domains/basic_radix2_domain.tcc#L33
+        if log_size_of_group > F::FftParams::TWO_ADICITY {
+            return None;
+        }
+
+        // Compute the generator for the multiplicative subgroup.
+        // It should be 2^(log_size_of_group) root of unity.
+        let group_gen = F::get_root_of_unity(num_coeffs)?;
+        let size_as_field_element = F::from(size);
+        let size_inv = size_as_field_element.inverse()?;
+
+        Some(Radix2EvaluationDomain {
+            size,
+            log_size_of_group,
+            size_as_field_element,
+            size_inv,
+            group_gen,
+            group_gen_inv: group_gen.inverse()?,
+            generator_inv: F::multiplicative_generator().inverse()?,
+        })
+    }
+
+    fn compute_size_of_domain(num_coeffs: usize) -> Option<usize> {
+        let size = num_coeffs.next_power_of_two();
+        // TODO: Check > vs. >= here.
+        if size.trailing_zeros() <= F::FftParams::TWO_ADICITY {
+            Some(size)
+        } else {
+            None
+        }
```

Comment:
I think it's correct to have `<=`, but let's make it consistent with the equivalent line in `new`, so that we use `>` in both places:

```suggestion
        let size = num_coeffs.next_power_of_two();
        if size.trailing_zeros() > F::FftParams::TWO_ADICITY {
            None
        } else {
            Some(size)
        }
```

Reply:
Good point, also I forgot to remove the TODO here. :)